src/core/net/ip6.cpp

```diff
@@ -504,6 +548,8 @@ ThreadError Ip6::HandleExtensionHeaders(Message &message, Header &header, uint8_
             break;
 
         case kProtoIp6:
+            ExitNow();
```

Comment:
Not quite got this one. After function exits here, it will call HandleDatagram() again as you pointed in the HandleDatagram() function since nextHeader will be set to kProtoIp6 (line 728).